kv/memberlist/memberlist_client.go

```diff
@@ -533,43 +529,83 @@ func (m *KV) JoinMembers(members []string) (int, error) {
 	return m.memberlist.Join(members)
 }
 
-func (m *KV) joinMembersOnStartup(ctx context.Context, members []string) error {
-	reached, err := m.memberlist.Join(members)
-	if err == nil {
-		level.Info(m.logger).Log("msg", "joined memberlist cluster", "reached_nodes", reached)
-		return nil
+// fastJoinMembersOnStartup attempts to reach small subset of nodes (computed as RetransmitMult * log10(number of discovered members + 1)).
+func (m *KV) fastJoinMembersOnStartup(ctx context.Context) {
+	startTime := time.Now()
+
+	members := m.discoverMembers(ctx, m.cfg.JoinMembers)
+	toReach := m.cfg.RetransmitMult * int(math.Ceil(math.Log10(float64(len(members)+1))))
+	if toReach > len(members) {
+		toReach = len(members)
 	}
 
-	if m.cfg.MaxJoinRetries <= 0 {
-		return err
+	level.Info(m.logger).Log("msg", "memberlist fast-join starting", "members_found", len(members), "to_reach", toReach)
+
+	totalReached := 0
+	previouslyAttemptedMembers := map[string]bool{}
+
+	for ; toReach > 0; members = m.discoverMembers(ctx, m.cfg.JoinMembers) {
+		// Remove members we have tried before.
+		for ix := len(members) - 1; ix >= 0; ix-- {
+			if previouslyAttemptedMembers[members[ix]] {
+				members = append(members[:ix], members[ix+1:]...)
+			}
+		}
+
+		// If there are no more nodes to try, we give up.
+		if len(members) == 0 {
+			break
+		}
+
+		if len(members) > toReach {
+			members = members[:toReach]
```

Comment:
I would randomize the list of members before selecting the subset. If `discoverMembers()` reply is consistent over time, this way different nodes starting up at nearly the same time will try to join a different set of members.